File: practices/chapter10_search/python/ch_10_search.py
import math

def seqFind(source=[], target=math.inf):
  for data in source:
    if data == target:
      return target
  return None

def get_middle_idx(_first_idx=0, _last_idx=0):
  return (_first_idx + _last_idx) // 2

# def mapFn(source=[]):
#   res = list(source)
#   for i in range(len(source)):
#     res[i] = {
#       'index': i,
#       'data': source[i]
#     }
#   return res

def binary_search(source=[], target=math.inf):
  # mapped_source = mapFn(source)
  # print(mapped_source)
  sorted_source = sorted(source)
  # print(sorted_source)
  
  first_idx = 0
  last_idx = len(source) - 1
  middle_idx = get_middle_idx(first_idx + last_idx)

  find_index = -1
  find_res = None
  find_times = 0

  while True:
    middle_data = sorted_source[middle_idx]
    find_times += 1
    if target == middle_data:
      find_res = target
      find_index = middle_idx
      break
    elif target < middle_data:
      last_idx = middle_idx - 1
    else:
      first_idx = middle_idx + 1
    middle_idx = get_middle_idx(first_idx, last_idx)
    if first_idx > last_idx:
      break

  return {
    'find_index': find_index,
    'find_res': find_res,
    'find_times': find_times,
  }

def find_max(source=[]):
  res = math.inf * -1
  for data in source:
    if data > res:
      res = data
  return res

def find_min(source=[]):
  res = math.inf
  for data in source:
    if data < res:
      res = data
  return res

def main():
  data_list = [9, 3, 2, 8, 7, 10, 6, 5, 4, 1,]
  target = eval(input('Input value you want to find: '))
  print('Search result(seq): ', seqFind(data_list, target))
  print('Search result(binary): ', binary_search(data_list, target))

  print('Min value: ', find_min(data_list))
  print('Max value: ', find_max(data_list))

if __name__ == "__main__":
    main()

File: practices/chapter10_search/js/ch_10_search.js
const {
  rl
} = require('../../lib/cmdReadline')
const {
  mergeSort
} = require('../../chapter9_sort/js/ch_9_merge_sort')


function seqFind(source=[], target=0) {
  let res = undefined

  source.forEach(data => {
    if(data === target) {
      res = target
    }
  })

  return res
}

function getMiddleIdx(idx1=0, idx2=0) {
  return Math.floor((idx1 + idx2) / 2)
}

function binaryFind(source=[], target=0) {
  const sorted = mergeSort(source)
  // console.log(sorted)

  let firstIdx = 0
  let lastIdx = source.length - 1
  let middleIdx = getMiddleIdx(firstIdx, lastIdx)

  let res = {
    index: -1,
    result: undefined,
    searchedTimes: 0,
  }

  while(true) {
    const middleData = sorted[middleIdx]
    res.searchedTimes += 1

    if(middleData === target) {
      res.index = middleIdx
      res.result = target
      break
    }
    else if(target > middleData) {
      firstIdx = middleIdx + 1
    }
    else {
      lastIdx = middleIdx - 1
    }
    // update middle index
    middleIdx = getMiddleIdx(firstIdx, lastIdx)

    if(firstIdx > lastIdx) {
      break
    }
  }

  return res
}

function findMin(source=[]) {
  let res = Infinity
  source.forEach(data => {
    if(data < res) {
      res = data
    }
  })
  return res
}

function findMax(source=[]) {
  let res = -Infinity
  source.forEach(data => {
    if(data > res) {
      res = data
    }
  })
  return res
}

function main() {
  const dataList = [9, 3, 2, 8, 7, 10, 6, 5, 4, 1,]
  rl.question('Input value you want to find: ', target => {
    console.log('Search result(seq): ', seqFind(dataList, Number(target)))
    console.log('Search result(binary): ', binaryFind(dataList, Number(target)))

    console.log('Min value: ', findMin(dataList))
    console.log('Max value: ', findMax(dataList))

    rl.close()
  })
}

main()